sdk/core/azure-core/src/main/java/com/azure/core/http/HttpRequest.java

```diff
@@ -278,4 +288,80 @@ public HttpRequest copy() {
+     * @return The request.
+     */
+    public HttpRequest copyWithRetryableBody() {
+        try {
+            final HttpHeaders bufferedHeaders = new HttpHeaders(headers);
+            BinaryData retryableBody = getRetryableBody();
+            return new HttpRequest(httpMethod, url, bufferedHeaders, retryableBody);
+        } finally {
+            requestBodyCopied = true;
+        }
+    }
+
+    private BinaryData getRetryableBody() {
+        if (body == null) {
+            return null;
+        }
+
+        BinaryDataContent content = BinaryDataHelper.getContent(body);
+
+        if (content instanceof ByteArrayContent
+            || content instanceof SerializableContent
+            || content instanceof StringContent
+            || content instanceof FileContent) {
+            return body;
+        } else if (content instanceof InputStreamContent) {
+            InputStream inputStream = content.toStream();
+            Long contentLength = getContentLength();
+            if (contentLength != null && contentLength < Integer.MAX_VALUE) {
+                if (inputStream.markSupported()) {
+                    if (requestBodyCopied) {
+                        try {
+                            inputStream.reset();
+                        } catch (IOException e) {
+                            throw LOGGER.logExceptionAsError(
+                                new UncheckedIOException("Unable to reset request body", e));
+                        }
+                    }
+                    inputStream.mark(contentLength.intValue());
+                    return body;
+                } else {
+                    // If stream is not seekable buffer body and set it in both derived and original request.
+                    byte[] bufferedContent = content.toBytes();
+                    BinaryData bufferedBinaryData = BinaryData.fromBytes(bufferedContent);
+                    this.setBody(bufferedBinaryData);
+                    return bufferedBinaryData;
+                }
+            } else {
+                throw LOGGER.logExceptionAsError(new IllegalStateException("Non retryable request body"));
+            }
+        } else {
+            /*
+             Clone the original request to ensure that each try starts with the original (unmutated) request. We cannot
+             simply call httpRequest.buffer() because although the body will start emitting from the beginning of the
+             stream, the buffers that were emitted will have already been consumed (their position set to their limit),
+             so it is not a true reset. By adding the map function, we ensure that anything which consumes the
+             ByteBuffers downstream will only actually consume a duplicate so the original is preserved. This only
+             duplicates the ByteBuffer object, not the underlying data.
+             */
+            Flux<ByteBuffer> bufferedBody = body.toFluxByteBuffer().map(ByteBuffer::duplicate);
```

Comment:
I agree, it's not bullet proof, especially if re-subscription won't produce same sequence... or the case you described.
Btw. deep copying buffer won't help either. When re-subscription happens both (original and deep-copied) would be gone. Unless... we preemptively buffer any flux - but that'd be excessive on sunny day scenario.
However, this change aims to pull existing stuff into central place from pieces scattered around like. 
https://github.com/Azure/azure-sdk-for-java/blob/97a86b9d2cf35043a55083c19574d8d8bd765967/sdk/storage/azure-storage-common/src/main/java/com/azure/storage/common/policy/RequestRetryPolicy.java#L97-L100
and
https://github.com/Azure/azure-sdk-for-java/blob/97a86b9d2cf35043a55083c19574d8d8bd765967/sdk/containerregistry/azure-containers-containerregistry/src/main/java/com/azure/containers/containerregistry/implementation/authentication/ContainerRegistryCredentialsPolicy.java#L94-L97

FWIW. storage scenario would work given this..
https://github.com/Azure/azure-sdk-for-java/blob/97a86b9d2cf35043a55083c19574d8d8bd765967/sdk/core/azure-core-http-netty/src/main/java/com/azure/core/http/netty/implementation/NettyAsyncHttpResponse.java#L38

Reply:
Unfortunately there's no good way to examine if `Flux<ByteBuffer>` is safe to retry and whether `ByteBuffer.duplicate` or `ByteBuffer.allocate()` (and deep copy) is enough. Some fluxes get away with just duplicating buffers, but some might not repeat same sequence upon resubscription. This is a good manifestation of the problem: https://github.com/Azure/azure-sdk-for-java/blob/97a86b9d2cf35043a55083c19574d8d8bd765967/sdk/storage/azure-storage-common/src/main/java/com/azure/storage/common/policy/RequestRetryPolicy.java#L170-L176

After a bit of debate with @alzimmermsft we figured that we'll manifest imperfection if this API `HttpRequest.copyWithRetryableBody(HttpRequestBodyBufferingMode bufferingMode)`/`HttpRequest.copyWithRetryableBodyAsync(HttpRequestBodyBufferingMode bufferingMode)` by adding mandatory parameter (enum flag) that let choose "buffering behavior" explicitly, so that caller may choose best option and is aware of what's happening. I.e.:
- DEEP - always buffer flux/stream into byte array
- SHALLOW - use mark/reset for inputstream, ByteBuffer.duplicate for fluxes
- NEVER - don't do anything.


@JonathanGiles @srnagar @lmolkova please chime in here and share your thoughts wrt. approach and naming. There's no perfect solution here, but we want to graduate from "some of this logic being scattered across sdks" to "have central place for this".